sway/input/keyboard.c

```diff
@@ -494,6 +514,52 @@ static void handle_keyboard_group_key(struct wl_listener *listener,
 	handle_key_event(sway_group->seat_device->keyboard, data);
 }
 
+static void handle_keyboard_group_enter(struct wl_listener *listener,
+		void *data) {
+	struct sway_keyboard_group *sway_group =
+		wl_container_of(listener, sway_group, enter);
+	struct sway_keyboard *keyboard = sway_group->seat_device->keyboard;
+	struct wl_array *keycodes = data;
+
+	uint32_t *keycode;
+	wl_array_for_each(keycode, keycodes) {
+		struct key_info keyinfo;
+		update_keyboard_state(keyboard, *keycode, WLR_KEY_PRESSED, &keyinfo);
+	}
+}
+
+static void handle_keyboard_group_leave(struct wl_listener *listener,
+		void *data) {
+	struct sway_keyboard_group *sway_group =
+		wl_container_of(listener, sway_group, leave);
+	struct sway_keyboard *keyboard = sway_group->seat_device->keyboard;
+	struct wl_array *keycodes = data;
+
+	uint32_t *keycode;
+	wl_array_for_each(keycode, keycodes) {
+		struct key_info keyinfo;
+		update_keyboard_state(keyboard, *keycode, WLR_KEY_RELEASED, &keyinfo);
+
+		bool pressed_sent = update_shortcut_state(&keyboard->state_pressed_sent,
+				*keycode, WLR_KEY_RELEASED, keyinfo.keycode, 0);
+		if (!pressed_sent) {
+			continue;
+		}
+
+		struct wlr_seat *wlr_seat = keyboard->seat_device->sway_seat->wlr_seat;
+		struct wlr_input_device *wlr_device =
+			keyboard->seat_device->input_device->wlr_device;
+
+		struct timespec now;
+		clock_gettime(CLOCK_MONOTONIC, &now);
+		int64_t time_msec = (int64_t)now.tv_sec * 1000 + now.tv_nsec / 1000000;
```

Comment:
I noticed we were going to duplicate this logic that `input/cursor.c` also has, so I refactored it in https://github.com/swaywm/sway/pull/5425. If that PR ends up getting merged, it'd remove the need to do this here as well.

Reply:
Sounds good. I didn't end up needing this after the changes, but it may come in handy in the future